drivers/io/io_block.c

```diff
@@ -0,0 +1,298 @@
+ */
+
+#include <assert.h>
+#include <debug.h>
+#include <errno.h>
+#include <io_block.h>
+#include <io_driver.h>
+#include <io_storage.h>
+#include <string.h>
+
+#define MIN_BLOCK_SIZE			64
+#define MIN_BLOCK_MASK			(MIN_BLOCK_SIZE - 1)
+
+io_type_t device_type_block(void);
+
+static int block_open(io_dev_info_t *dev_info, const uintptr_t spec,
+		      io_entity_t *entity);
+static int block_seek(io_entity_t *entity, int mode, ssize_t offset);
+static int block_read(io_entity_t *entity, uintptr_t buffer, size_t length,
+		      size_t *length_read);
+static int block_write(io_entity_t *entity, uintptr_t buffer, size_t length,
+		       size_t *length_written);
+static int block_close(io_entity_t *entity);
+static int block_dev_init(io_dev_info_t *dev_info,
+			  const uintptr_t init_params);
+static int block_dev_open(const uintptr_t dev_spec, io_dev_info_t **dev_info);
+static int block_dev_close(io_dev_info_t *dev_info);
+
+static const io_dev_connector_t block_dev_connector = {
+	.dev_open	= block_dev_open
+};
+
+static const io_dev_funcs_t block_dev_funcs = {
+	.type		= device_type_block,
+	.open		= block_open,
+	.seek		= block_seek,
+	.size		= NULL,
+	.read		= block_read,
+	.write		= block_write,
+	.close		= block_close,
+	.dev_init	= block_dev_init,
+	.dev_close	= block_dev_close,
+};
+
+static io_dev_info_t block_dev_info = {
+	.funcs	= &block_dev_funcs,
+	.info	= (uintptr_t)NULL,
+};
+
+io_type_t device_type_block(void)
+{
+	return IO_TYPE_BLOCK;
+}
+
+static int block_open(io_dev_info_t *dev_info, const uintptr_t spec,
+		      io_entity_t *entity)
+{
+	io_block_entity_t *block;
+	io_block_spec_t *region;
+
+	assert((dev_info->info != (uintptr_t)NULL) &&
+	       (spec != (uintptr_t)NULL) &&
+	       (entity->info == (uintptr_t)NULL));
+
+	block = (io_block_entity_t *)spec;
+	region = &block->region;
+	memcpy((void *)&block->buffer, (void *)dev_info->info,
+	       sizeof(io_block_spec_t));
+
+	assert((block->block_size > 0) &&
+	       ((block->block_size & MIN_BLOCK_MASK) == 0) &&
+	       ((region->offset % block->block_size) == 0) &&
+	       ((region->length % block->block_size) == 0) &&
+	       ((block->buffer.offset % block->block_size) == 0) &&
+	       ((block->buffer.length % block->block_size) == 0));
+
+	block->file_pos = 0;
+	block->lba = (block->file_pos + region->offset) / block->block_size;
+	entity->info = spec;
+	return 0;
+}
+
+/* parameter offset is relative address at here */
+static int block_seek(io_entity_t *entity, int mode, ssize_t offset)
+{
+	io_block_entity_t *block;
+	io_block_spec_t *region;
+
+	assert(entity->info != (uintptr_t)NULL);
+
+	block = (io_block_entity_t *)entity->info;
+	region = &block->region;
+
+	assert(offset < block->region.length);
+
+	block->file_pos += offset;
+	block->lba = (block->file_pos + region->offset) / block->block_size;
+	assert(block->lba >= 0);
+	return 0;
+}
+
+static int block_read(io_entity_t *entity, uintptr_t buffer, size_t length,
+		      size_t *length_read)
```

Comment:
Both the read and write functions could do with more comments or a big comment for each explaining what is going on.